source/agora/common/Amount.d

```diff
@@ -0,0 +1,239 @@
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.Amount;
+
+/// Ditto
+public struct Amount
+{
+    import agora.common.Deserializer;
+    import agora.common.Hash;
+    import agora.common.Serializer;
+
+    import core.checkedint;
+
+    /// Number of units ('cents') per coins
+    public static immutable Amount UnitPerCoin   = Amount(10_000_000, true);
+    /// Maximum number of BOA coins that can ever be in circulation
+    public static immutable ulong  MaxCoinSupply = 500_000_000;
+    /// Maximum amount of money that can ever be in circulation
+    public static immutable Amount MaxUnitSupply =
+        Amount(UnitPerCoin.value * MaxCoinSupply, true);
+
+    /// Internal data storage
+    private ulong value;
+
+    /// Support for hashing
+    public void computeHash(scope HashDg dg) const nothrow @nogc @safe
+    {
+        hashPart(this.value, dg);
+    }
+
+    /// Support for serialization
+    public void serialize(scope SerializeDg dg) const nothrow @safe
+    {
+        serializePart(this.value, dg);
+    }
+
+    /// Support for deserialization
+    public void deserialize(scope DeserializeDg dg) nothrow @safe
+    {
+        deserializePart(this.value, dg);
+    }
+
+    /// Pretty-print this value
+    public void toString (SinkT dg) const @safe
+    {
+        import std.format;
+        formattedWrite(dg, "%d", this.value);
+    }
+
+    /// Also support for Vibe.d serialization to JSON
+    public string toString () const @safe
+    {
+        string ret;
+        scope SinkT dg = (scope v) { ret ~= v; };
+        this.toString(dg);
+        return ret;
+    }
+
+    private alias SinkT = void delegate(scope const(char)[] v) @safe;
+
+    /// Support for Vibe.d deserialization
+    public static Amount fromString (scope const(char)[] str) pure @safe
+    {
+        import std.conv : to;
+        immutable ul = str.to!ulong;
+        if (!Amount.isInRange(ul))
+            throw new Exception("Invalid input value to Amount");
+        return Amount(ul, true);
+    }
+
+
+    nothrow pure @nogc @safe:
+
+    /***************************************************************************
+
+        Construct an instance of an `Amount`
+
+        Params:
+            units = The monetary amount this new instance represent.
+                    The unit is the smaller unit of currency (see `UnitPerCoin`)
+
+    ***************************************************************************/
+
+    public this (ulong units)
+    {
+        assert(Amount.isInRange(units));
+        this.value = units;
+    }
+
+    /// Make a value without checking bounds
+    /// Used to initialize the bounds themselves, and to make invalid values
+    /// in unittests
+    private this (ulong units, bool dummy) { this.value = units; }
```

Comment:
```suggestion
        private this (ulong units, Flag!"YouDingus") { this.value = units; }
```

And then use it like:

```D
auto amount = Amount(100, Yes.YouDingus);
```

Reply:
You have terrible indentation tastes

Reply:
Whoa what happened there. Ignore that please.

Reply:
Too late. Can't unsee.